drivers/retained_mem/retained_mem_zephyr_ram.c

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2023, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_retained_ram
+
+#include <string.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/retained_mem.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(retained_mem_zephyr_ram, CONFIG_RETAINED_MEM_LOG_LEVEL);
+
+#ifdef CONFIG_MULTITHREADING
+struct zephyr_retained_mem_ram_data {
+	struct k_mutex lock;
+};
+#endif
+
+struct zephyr_retained_mem_ram_config {
+	uint8_t *address;
+	size_t size;
+};
+
+static inline void zephyr_retained_mem_ram_lock_take(const struct device *dev)
+{
+#ifdef CONFIG_MULTITHREADING
+	struct zephyr_retained_mem_ram_data *data = dev->data;
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+#else
+	ARG_UNUSED(dev);
+#endif
+}
+
+static inline void zephyr_retained_mem_ram_lock_release(const struct device *dev)
+{
+#ifdef CONFIG_MULTITHREADING
+	struct zephyr_retained_mem_ram_data *data = dev->data;
+
+	k_mutex_unlock(&data->lock);
+#else
+	ARG_UNUSED(dev);
+#endif
+}
+
+static int zephyr_retained_mem_ram_init(const struct device *dev)
+{
+#ifdef CONFIG_MULTITHREADING
+	struct zephyr_retained_mem_ram_data *data = dev->data;
+
+	k_mutex_init(&data->lock);
+#endif
+
+	return 0;
+}
+
+static ssize_t zephyr_retained_mem_ram_size(const struct device *dev)
+{
+	const struct zephyr_retained_mem_ram_config *config = dev->config;
+
+	return (ssize_t)config->size;
+}
+
+static int zephyr_retained_mem_ram_read(const struct device *dev, size_t offset, uint8_t *buffer,
+					size_t size)
+{
+	const struct zephyr_retained_mem_ram_config *config = dev->config;
+
+	zephyr_retained_mem_ram_lock_take(dev);
+
+	memcpy(buffer, (config->address + offset), size);
```

Comment:
Before `memcpy` is called, the `config->address + offset` and `size` should be checked if they are within allowed range of the allowed "retained ram", otherwise you could use the call to write anywhere within address space.

Reply:
My bad, it is checked in syscall.